### PR TITLE
feat: support state destructure

### PR DESCRIPTION
**Pull request by decadef20**
## Description

Add a short description of:
- what changes you made, 
 #389 
- why you made them, and 
- any other context that you think might be helpful for someone to better understand what is contained in this pull request. 

This sort of information is useful for people reviewing the code, as well as anyone from the future trying to understand why changes were made or why a bug started happening.

### CODE REVIEW

File: packages/core/src/__tests__/__snapshots__/react.test.ts.snap [hunk 1]
```diff
@@ -1610,3 +1610,28 @@ foo: string
...
+exports[`React stateDestructure 1`] = `
+"import { useState, useEffect } from \\"react\\";
+
+export default function MyBasicComponent(props) {
+  const [name, setName] = useState(() => \\"Decadef20\\");
+
+  const [sports, setSports] = useState(() => []);
+
+  useEffect(() => {
+    const name = name,
```
Comment (decadef20) [inline packages/core/src/__tests__/__snapshots__/react.test.ts.snap:1623]: The result is weird.  
Should we support `const { name } = state` ?

Comment (decadef20) [reply packages/core/src/__tests__/__snapshots__/react.test.ts.snap:1623]: I think this line should be deleted in react framework.

File: packages/core/src/__tests__/data/basic-state.raw.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  });
+
+  onMount(() => {
+    const n = state.name,
+      sports = state.sports;
+  });
+
+  onInit(() => {
+    const name = state.name,
+      sports = state.sports,
```
[the remark this answers is not in the document]
Comment (samijaber) [reply packages/core/src/__tests__/data/basic-state.raw.tsx:16]: The issue is that a bunch of Mitosis frameworks have their props and state at the module scope, like Svelte and Vue. So the props, state, and constants might share scope, which can become problematic.

@decadef20 Will need some time to think about the implications of this PR before I can properly review.

[the remark this answers is not in the document]
Comment (decadef20) [reply packages/core/src/__tests__/data/basic-state.raw.tsx:16]: >https://github.com/BuilderIO/mitosis/blob/main/packages/eslint-plugin/docs/rules/no-var-name-same-as-prop-name.md#L1
https://github.com/BuilderIO/mitosis/blob/main/packages/eslint-plugin/docs/rules/no-var-name-same-as-state-property.md#L1

Very clear doc.   

That's to say, should we support state destructuring? If not, this PR isn't needed and should be closed. If so, it seems a long way to go

[the remark this answers is not in the document]
Comment (PatrickJS) [reply packages/core/src/__tests__/data/basic-state.raw.tsx:16]: @samijaber maybe we should support this as undocumented for now until we figure out the edge case? what do you think? or just close it

[the remark this answers is not in the document]
Comment (samijaber) [reply packages/core/src/__tests__/data/basic-state.raw.tsx:16]: I would avoid supporting for now, and we can re-consider down the line when it comes up again 👍🏽

**Pull request closed without merging**